rpc/src/v1/helpers/errors.rs

```diff
@@ -422,6 +422,14 @@ pub fn filter_not_found() -> Error {
 	}
 }
 
+pub fn filter_block_not_found() -> Error {
+	Error {
+		code: ErrorCode::ServerError(codes::UNSUPPORTED_REQUEST),
+		message: "One of the block specified in filter (fromBlock, toBlock or blockHash) cannot be found".into(),
+		data: None,
```

Comment:
EIP234 requires that `data` include the hash of the requested block. Since we're also returning an error on `to/from` what do you think we should do? Is it really useful to return the requested hash if not found?

Reply:
I think it's equally useful to return the `from/to` block, since there're two and the caller might want to know which. Let me fix this.

Reply:
A few thoughts on this:
1. If a user asks for blocks `6,000,000` to `7,000,000` but the highest block is `6,500,000`, should we interpret that to mean the user wants "all blocks from `6,000,000` to latest"?
2. If we consider it an error to ask for more blocks than latest, can the error data be the lowest block number that wasn't present?  So in the above example the error data would be `6,500,001`?

I haven't put a whole lot of thought into the above, @epheph may have some feelings since he has experienced a lot of the pain around this.  My gut tells me that it should return an error because otherwise the user doesn't know which blocks had no logs and which blocks were missing, so it doesn't know what to ask for again.  This is especially problematic when dealing with load balanced servers like Infura where each request may be routed differently and thus requests may go to a server that is behind (thus missing blocks that another had near head).

Typing all that out makes me reasonably confident that an error is in fact correct behavior.  In that case, letting the user know what blocks are missing (rather than just "some blocks in the range") could prove useful to them, especially if they are attempting to fetch a very large range.

Reply:
@MicahZoltu Yes, the current behavior would return error with the error data pointing to block `7,000,000`. :)

I think returning `6,500,001` may be too costly on the client side. It needs to (at least partially) iterate over the block range.

Reply:
An error when your range exceeds the available blocks, even by a single block, is the best action. Returning that `6,500,001` is where that cut-off is would be nice for troubleshooting (and more correct), but not an absolutely requirement. None of the applications I've worked on would have taken any other action besides re-requesting the entire batch that was originally requested.

(It's possible the app could read up to the available block and then wait for more to become available, but requesting large ranges of blocks is usually done for blocks significantly in the past. I believe this error will occur most when you are asking for small numbers of blocks very close to head)

Reply:
Yeah like you said, I can't think of a use-case where we may query a non-existent `fromBlock` or `toBlock`. If a query doesn't care about it, it can just set `toBlock` to `latest` or `null`.

Reply:
We have cases where we query potentially-non-existent blocks. When using a pool of eth nodes (such as Infura), one might receive a websocket notification of newHead `block X` (or when getBlockByNumber("latest") ), then in a follow-up query to getLogs from that block, it might hit a node that does not yet have that block. In this case, `latest` probably isn't good because `latest` is likely to have changed. My view of the purpose of EIP-234 is to remove the race condition that exists between a blockNumber and a blockHash.

I don't mind getting an error back if the block doesn't exist, i just don't want to get back something I can't distinguish from a successful fetch of no logs.